llvm/include/llvm/SandboxIR/Context.h

```diff
@@ -48,6 +64,21 @@ class Context {
   /// Type objects.
   DenseMap<llvm::Type *, std::unique_ptr<Type, TypeDeleter>> LLVMTypeToTypeMap;
 
+  /// Callbacks called when an IR instruction is about to get removed. Keys are
+  /// used as IDs for deregistration.
+  DenseMap<CallbackID, RemoveInstrCallback> RemoveInstrCallbacks;
+  /// Callbacks called when an IR instruction is about to get inserted. Keys are
+  /// used as IDs for deregistration.
+  DenseMap<CallbackID, InsertInstrCallback> InsertInstrCallbacks;
+  /// Callbacks called when an IR instruction is about to get moved. Keys are
+  /// used as IDs for deregistration.
+  DenseMap<CallbackID, MoveInstrCallback> MoveInstrCallbacks;
+
+  /// A counter used for assigning callback IDs during registration. The same
+  /// counter is used for all kinds of callbacks so we can detect mismatched
+  /// registration/deregistration.
+  static CallbackID NextCallbackID;
```

Comment:
this shouldn't be static?

Reply:
Depends on whether we want callback IDs to be unique across the whole run or just for a single context object. I figured unique-per-run might be good for debugging (you never get two callback registrations with the same ID), but in practice I don't think it will make a big difference, if at all.

Reply:
I believe people will run multiple LLVM pipelines in parallel in one process, although with different LLVMContexts. so anything static will unintentionally be shared between those threads and be subject to data races

Reply:
Ok, that's a good reason. Made it non-static.